Bug 1346567 - cycle_data is failing on stage/production

ionutgoldan opened the pull request.

ionutgoldan commented on treeherder/perf/models.py:
```diff
@@ -107,32 +108,94 @@ class PerformanceDatumManager(models.Manager):
+
+                try:
+                    cursor.execute('''
+                        DELETE FROM `performance_datum`
```
@sarah-clements I've decided to go for a raw query, because Django's ORM isn't capable of building DELETE w/ LIMIT queries.

Using the same cycling algorithm like that from Treeherder cannot delete more than 4000 rows/minute, which is way too slow for us. We would need 60 days just to be sure we keep Perfherder data under 500 billion rows.
From what I noticed, it's because we cannot simply increase the chunk_size to more than a few hundreds, when doing this; somehow, the query begins to freeze, then times out.

Anyway, the raw query seems to behave better with chunk_sizes of 10,000, but I need to test this 1st on treeherder-prototype.

ionutgoldan replied:
> I was talking to @klahnakoski about this and he thinks a good approach would be to set it to delete 1000 rows at a time (or the upper threshold of say 4000). It may indeed take a while to delete the old data but this way it will continuously delete the data at a rate that shouldn't cause issues (like a lock)- say every minute or 5 minutes.

You mean schedule the script to run every 1 or 5 minutes & only delete 1000 rows each time, then close?
This sounds like a safe solution & we should have considerable fewer lock issues.
Still, Heroku Scheduler can run every 10 minutes/1 hour/ day. This means we'll delete 144,000 rows (for chunk size at 1000) or 576,000 rows (for chunk size at 4000).

klahnakoski replied:
Maybe a tight loop: inside is a transaction that will delete up to 1000 rows.  This will 
* delete more rows in a single run, 
* it can be simply scheduled once per day
* it will exit when there is nothing left to delete
* if it crashes then it will at least have done some work
* the small transactions will allow all other work to proceed as usual

klahnakoski commented on treeherder/perf/models.py:
```diff
@@ -107,32 +108,94 @@ class PerformanceDatumManager(models.Manager):
+                try:
+                    cursor.execute('''
+                        DELETE FROM `performance_datum`
+                        WHERE repository_id = %s AND
```
Does the `repository_id` matter?  Can all repos be done at once?  If so, this can run faster, and leave less holes in the table.

I am also concerned that a full table scan is required if the `chunk_size > number_of_records_to_delete`, which will then timeout.  Maybe a rough upper bound on the `id`can help with that:
    
    select id into @max_id from performance_datum where push_timestamp>'2018-07-22' limit 1;
    select id from performance_datum where push_timestamp<'2018-07-22' and id<@max_id limit 1000;

ionutgoldan replied:
> Does the `repository_id` matter? Can all repos be done at once? If so, this can run faster, and leave less holes in the table.

Not really, we could delete rows without it.

> I am also concerned that a full table scan is required if the `chunk_size > number_of_records_to_delete`, which will then timeout. Maybe a rough upper bound on the `id`can help with that:
> 
> ```
> select id into @max_id from performance_datum where push_timestamp>'2018-07-22' limit 1;
> select id from performance_datum where push_timestamp<'2018-07-22' and id<@max_id limit 1000;
> ```

@klahnakoski I guess it's best to compute this upper bound right before running the DELETE ... LIMIT query.

klahnakoski commented on treeherder/perf/models.py:
```diff
@@ -107,32 +108,94 @@ class PerformanceDatumManager(models.Manager):
+
+            if sleep_time:
+                # Allow some time for other queries to get through
+                time.sleep(sleep_time)
```
There is no need to sleep.  Assuming each `execute()` is inside its own transaction:  Other work will be queued by the database while `execute()` runs.  When `execute()` is done, the other work will be run before the next `execute()`.

ionutgoldan commented:
I've addressed your suggestions in the 2nd commit. I noticed the related tests no longer work, but will fix them next thing in the morning.

ionutgoldan commented:
I've deployed this on treeherder-prototype & setup a new daily Heroku scheduler worker. It no longer experienced timeouts.
I estimate it's able to delete 140 million rows per day.
I'll do some more testing of this PR tomorrow.

ionutgoldan commented:
I've deployed this on treeherder-prototype & setup a new daily Heroku scheduler worker. It no longer experienced timeouts.
I estimate it's able to delete 140 million rows per day.
I'll do some more testing of this PR tomorrow.

ionutgoldan commented:
@sarah-clements The scheduler finished running. It managed to delete around 190 million rows from `performance_datum` alone :). So even more than my initial estimation.

This means the updated PR is ready for review.

sarah-clements commented on treeherder/perf/models.py:
```diff
@@ -143,34 +131,35 @@ def cycle_data(self, repository, cycle_interval, chunk_size, sleep_time,
+                        WHERE push_timestamp < %s
                         LIMIT %s
-                    ''', [repository.id, max_timestamp, chunk_size])
+                    ''', [max_timestamp, ideal_chunk_size])
```
Could you you do something like this instead of the raw SQL? 
https://stackoverflow.com/questions/13248593/django-database-delete-specific-number-of-entries

ionutgoldan replied:
> Could you you do something like this instead of the raw SQL?
> https://stackoverflow.com/questions/13248593/django-database-delete-specific-number-of-entries

I tried to do that locally, but I get `django.db.utils.NotSupportedError: (1235, "This version of MySQL doesn't yet support 'LIMIT & IN/ALL/ANY/SOME subquery'")`, just like one of the Stackoverflow users reported.

armenzg replied:
What version do you need?
AWS allow us to upgrade to 8.0.15 and 5.7.26 (we're on 5.7.23)

sarah-clements commented on treeherder/perf/models.py:
```diff
@@ -108,33 +107,22 @@ class PerformanceDatumManager(models.Manager):
-        max_timestamp = self._determine_max_timestamp(repository.name,
-                                                      cycle_interval,
-                                                      unsheriffed_repos_expire_days)
+        logger.warning('Cycling Perfherder data...')
```
I think this would add more noise to papertrail and isn't as useful as the success and exceptions warnings.

ionutgoldan replied:
Agree.

sarah-clements commented on treeherder/model/management/commands/cycle_data.py:
```diff
@@ -82,40 +81,27 @@ class PerfherderCycler(DataCycler):
-
     def cycle(self):
         self.started_at = datetime.datetime.now()
+        max_chunk_size = self.chunk_size
```
Could you make these names the same for consistency?

ionutgoldan replied:
Sure thing.

sarah-clements commented on treeherder/perf/models.py:
```diff
@@ -183,18 +172,20 @@ def _delete_in_chunks(self, repository, max_timestamp, chunk_size, sleep_time, l
-    def _determine_max_timestamp(self, repo_name, cycle_interval, unsheriffed_repos_expire_days):
-        if repo_name in self.MAIN_REPOS:
-            return datetime.datetime.now() - cycle_interval
+    def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
```
I'm wondering why calculate the length of the older data versus just using the max_chunk_size?

klahnakoski replied:
Probably from my concern for the case when the number of records left to delete is less than the chunk size: The query will continue to with the table scan looking for more, and it will timeout doing so.  

The reason this is faster is because it is only an approximation: there will be some records with higher ids, but lower timestamps that could be deleted.  I assume that is relatively small

ionutgoldan replied:
I share Kyle's opinion on this.

sarah-clements replied:
Ok, makes sense.

sarah-clements commented on treeherder/model/management/commands/cycle_data.py:
```diff
@@ -82,40 +81,27 @@ class PerfherderCycler(DataCycler):
 
-            self.logger.warning('Performance data cycling lasted for {}h{}s'
-                                .format(hours, seconds))
+    def report_runtime(self):
```
Is this still useful? I think we'd be able to see how long it's been running from the Papertrail logs and based on what the start time is set in the Heroku scheduler.

ionutgoldan replied:
Yes, that's a good point.

sarah-clements commented on treeherder/perf/models.py:
```diff
@@ -143,34 +131,35 @@ def cycle_data(self, repository, cycle_interval, chunk_size, sleep_time,
-        for signature in PerformanceSignature.objects.filter(
-                repository=repository):
+        for signature in PerformanceSignature.objects.all():
+            if self._should_quit(started_at, max_overall_runtime):
```
I'm curious why time it and end it this way versus catching a timeout exception?

ionutgoldan replied:
> I'm curious why time it and end it this way versus catching a timeout exception?

Indeed, this would make the code more readable.

ionutgoldan commented:
I've addressed the 2nd round of review requests, excepting the Stackoverflow suggested alternative for the raw SQL query *(because it doesn't work in our case)*.

I added an extra check for the `--days` argument. If it's smaller than 365 days when cycling perf data, the CLI will fail fast. I did this because it's very easy to mistakenly set the wrong value when configuring the Heroku scheduler.
We really don't want to delete data that's more recent than a year.

ionutgoldan commented:
>What are the flags/params we need to add to the heroku scheduler?

./manage.py cycle_data --days 365 --chunk-size 1000 from:perfherder

sarah-clements commented on treeherder/model/management/commands/cycle_data.py:
```diff
@@ -81,27 +82,17 @@ class PerfherderCycler(DataCycler):
         super().__init__(days, chunk_size, sleep_time, is_debug, logger)
-        self.started_at = None
+        if days < MINIMUM_PERFHERDER_EXPIRE_INTERVAL:
+            raise ValueError('Cannot expiry too recent performance data')
```
Perhaps it'd be useful to show what the limit is by including the variable: "Cannot expire data earlier than <expire interval> days"?

ionutgoldan replied:
Yes, that'd be useful to know from the logs.

sarah-clements commented:
> I've addressed the 2nd round of review requests, excepting the Stackoverflow suggested alternative for the raw SQL query _(because it doesn't work in our case)_.
> 
> I added an extra check for the `--days` argument. If it's smaller than 365 days when cycling perf data, the CLI will fail fast. I did this because it's very easy to mistakenly set the wrong value when configuring the Heroku scheduler.
> We really don't want to delete perf data that's more recent than a year.
> 
Since this value shouldn't really change, perhaps it'd be better to set it as a default value in the django command or either the Perfherder class, with the option to pass a different value if in debug mode?

sarah-clements commented:
> > What are the flags/params we need to add to the heroku scheduler?
> 
> ./manage.py cycle_data --days 365 --chunk-size 1000 from:perfherder
> 
> Heroku Scheduler should run this on a daily basis.

And the only change to the treeherder cycle data command will be adding the `from:treeherder` param, right?

ionutgoldan commented:
> 
> 
> > > What are the flags/params we need to add to the heroku scheduler?
> > 
> > 
> > ./manage.py cycle_data --days 365 --chunk-size 1000 from:perfherder
> > Heroku Scheduler should run this on a daily basis.
> 
> And the only change to the treeherder cycle data command will be adding the `from:treeherder` param, right?

Not really. You can leave that Heroku Scheduler command unchanged. `from:treeherder` is the default, when no subcommand is specified.

ionutgoldan commented:
> 
> 
> > I added an extra check for the `--days` argument. If it's smaller than 365 days when cycling perf data, the CLI will fail fast. I did this because it's very easy to mistakenly set the wrong value when configuring the Heroku scheduler.
> > We really don't want to delete perf data that's more recent than a year.
> 
> Since this value shouldn't really change, it'd be preferable to set it as a default value in the Command class or the Perfherder class, with the option to pass a different value if in debug mode. So then the `--days 365` flag wouldn't be needed.

That's a fair point. I would still like to have it explicitly expressed in Heroku Scheduler's CLI, as a quick source of truth, in terms of config params.
Also, I will need to add another `days` argument for other test types. In a separate future PR, that is.
Having these both visible will quickly hint the config of the expiring algorithm.

ionutgoldan commented:
> Thanks for making the changes. Do these need to be tested on prototype before merging to stage?

I would prefer doing a small recheck on prototype. Run the script for 3-4 hours there.

armenzg commented:
I was curious about this and I peeked at the free storage for the DB instance:
<img width="1828" alt="image" src="https://user-images.githubusercontent.com/44410/62132604-595e1300-b2ab-11e9-85f9-5f61dab8d68e.png">

3-day storage:
<img width="683" alt="image" src="https://user-images.githubusercontent.com/44410/62132696-87435780-b2ab-11e9-849b-86ef7b257390.png">

I like what I see. Good job @ionutgoldan

klahnakoski commented:
@armenzg The chart is *free* storage space, and it is going down.  That means the amount stored is going up, and that seems bad.

armenzg commented:
> @armenzg The chart is _free_ storage space, and it is going down. That means the amount stored is going up, and that seems bad.

I understand that, however, there's quite the improvement from the runs from @ionutgoldan 
We will need to do further improvements in the future, however, the latest trend is more flat than what it has been.
We're monitoring free storage and we can cheaply increase it.

1 day
<img width="699" alt="image" src="https://user-images.githubusercontent.com/44410/62138917-7b10c780-b2b6-11e9-8983-19b318d7e5c4.png">
12 hours
<img width="704" alt="image" src="https://user-images.githubusercontent.com/44410/62138943-83690280-b2b6-11e9-85bc-dd765a5ff9ea.png">

sarah-clements commented:
> > Thanks for making the changes. Do these need to be tested on prototype before merging to stage?
> 
> I would prefer doing a small recheck on prototype. Run the script for 3-4 hours there.
> Then merge it to master & test on staging for the 3-4 days, when it will bring the database in a good state, with mostly recent data.

Sounds good!

ionutgoldan commented:
@camd could you also take a look over this PR?

ionutgoldan commented:
Tested this on `treeherder-prototype` for 23 hours & it didn't timeout. It almost removed all old data, except for one month worth of it.

armenzg commented:
In production.
